CDAP-6145 Docker Standalone Capabilities

awholegunch opened the pull request:
Fix for https://issues.cask.co/browse/CDAP-6145

Running as a Quick Build: http://builds.cask.co/browse/CDAP-DQB242-3

Page of interest: http://builds.cask.co/artifact/CDAP-DQB242/shared/build-3/Docs-HTML/4.1.0-SNAPSHOT/en/developers-manual/getting-started/standalone/docker.html

Commit:
work-in-progress

Commit:
Updates.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -86,7 +92,20 @@ started correctly.
+     :dependent: linux-windows
+     :languages: console,shell-session
+ 
+     $ docker run -it --name cdap-sdk-debugging -p 11011:11011 -p 11015:11015 caskdata/cdap-standalone:|release| cdap.sh start --enable-debug 
```
`cdap sdk start --enable-debug`

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +117,101 @@ started correctly.
-     
-     $ docker exec -d cdap-standalone /opt/cdap/sdk/bin/cdap sdk stop
-     $ docker-machine stop cdap
+    $ docker exec -it cdap-sdk cdap-cli.sh
```
`docker exec -it cdap-sdk cdap cli`

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +117,101 @@ started correctly.
-     > docker-machine stop cdap
+- Use the CDAP CLI in its own container (*cdap-cli*), against a remote CDAP instance at ``${CDAP_HOST}``::
+
+    $ docker run -it --name cdap-cli --rm caskdata/cdap-standalone cdap-cli.sh -u http://${CDAP_HOST}:11011
```
`docker run -it --name cdap-cli --rm caskdata/cdap-standalone cdap cli -u http://${CDAP_HOST}:11011`

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +117,101 @@ started correctly.
+
+- Use the CDAP CLI in its own container (*cdap-cli*), against the above *cdap-standalone* container using container linking::
+
+    $ docker run -it --link cdap-sdk:sdk --name cdap-cli --rm caskdata/cdap-standalone sh -c 'exec cdap-cli.sh -u http://${SDK_PORT_11011_TCP_ADDR}:${SDK_PORT_11011_TCP_PORT}'
```
s/cdap-cli.sh/cdap cli/

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +117,101 @@ started correctly.
+
+- Starting the Standalone CDAP, in the foreground, with ports forwarded::
+
+    $ docker run -it -p 11015:11015 -p 11011:11011 --name cdap-sdk caskdata/cdap-standalone cdap.sh start
```
s/cdap.sh/cdap sdk/

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +117,101 @@ started correctly.
+    
+- Starting the Standalone CDAP, in the foreground, with ports forwarded, and with debugging enabled::
+
+    $ docker run -it -p 11015:11015 -p 11011:11011 --name cdap-sdk caskdata/cdap-standalone cdap.sh start --enable-debug
```
s/cdap.sh/cdap sdk/

awholegunch replied:
fixed.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +117,101 @@ started correctly.
+  <https://docs.docker.com/engine/tutorials/dockervolumes/#/adding-a-data-volume>`__.
+  You can either let Docker manage the storage in the container (the easiest) or you can 
+  `mount a host directory as a data volume
+  <https://docs.docker.com/engine/tutorials/dockervolumes/#/mount-a-host-directory-as-a-data-volume>`.
```
Missing ending underscores?

awholegunch replied:
Fixed.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +117,101 @@ started correctly.
+    > docker exec -d cdap-standalone /opt/cdap/sdk/bin/cdap sdk stop
+
+    # To see the status:
+    > docker exec -it cdap-standalone /opt/cdap/sdk/bin/cdap sdk status
```
These are all `docker exec ${opts} ${container_name} cdap sdk ${command}` where:
- ${opts} is one of `-it` or `-d` for foreground/background
- ${container_name} is always `cdap-sdk` (--name from the `docker run`)
- `cdap` script is in the PATH, always... we can specify it, but it's unnecessary
- ${command} is the argument to `cdap sdk`

You can do check `status` of two ways, as you've documented, where you do it in the foreground (-it) and parse status with your eyes, or you can do it in the background like the others and check exit status, which is better for scripted use. I think this is fine for the user-oriented case, though.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +117,101 @@ started correctly.
+
+    .. Windows
+    
+    > docker exec -d cdap-standalone /opt/cdap/sdk/bin/cdap sdk stop
```
These should also be s/cdap-standalone/cdap-sdk/ and just remove the full PATH to `cdap sdk`

Commit:
Fixes per comments. Added a "keepslashes" option to the tabbed-parsed-literal
to preserve all slashes in a command for cases such as these.

awholegunch commented:
I have updated per comments. I added a new flag (`keepslashes`) to our custom tabbed-parsed-literal directive; this means that the same line can be used for both Linux and Windows, and will not have forward-slashes converted to backslashes. This will reduce the duplication of many lines.

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -51,30 +54,41 @@ started correctly.
+     :languages: console,shell-session
+     :keepslashes:
+ 
+     $ docker run -d --name cdap-standalone -p 11011:11011 -p 11015:11015 caskdata/cdap-standalone:|release|
```
I missed this, we should use `--name cdap-sdk` since that's what we're using below.

awholegunch replied:
Fixed

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -51,30 +54,41 @@ started correctly.
+ 
+     $ docker run -d --name cdap-standalone -p 11011:11011 -p 11015:11015 caskdata/cdap-standalone:|release|
+     
+   This will start the container (in the background), name it ``cdap-standalone``, and set the proxying of ports.
```
Use `cdap-sdk` here, too

awholegunch replied:
Fixed

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +114,123 @@ started correctly.
+    :languages: console,shell-session
+    :keepslashes:
+ 
+    $ docker run -it --name cdap-cli --rm caskdata/cdap-standalone cdap cli -u http://${CDAP_HOST}:11011
```
`11015`

awholegunch replied:
Fixed

wolf31o2 commented on cdap-docs/developers-manual/source/getting-started/standalone/docker.rst:
```diff
@@ -98,43 +114,123 @@ started correctly.
+  However, if you mount a host directory, you must make sure that it exists and that
+  permissions are set correctly. You pass such a directory using ``-v
+  /my/own/datadir:/opt/my/own/datadir``, where mounts the ``/my/own/datadir`` from the
+  host system as ``/opt/cdap/sdk/my/own/datadir`` in the container.
```
These PATHs don't match.

awholegunch replied:
Fixed.

wolf31o2 reviewed (changes requested):
Couple more minor changes...

Commit:
Fixes per comments.

wolf31o2 reviewed (approved):
👍 LGTM
